### DISCUSSION TITLE
[Feature]: CurvedText (HELP)

### DISCUSSION THREAD

**Discussion by vavilon224** [fabricjs/fabric.js#8861] [Q&A]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have searched and referenced existing issues, feature requests and discussions
- [X] I am filing a **FEATURE** request.

### Description

Are there any plans to add a curved text feature? I see there has been a lot of discussion over the years on Stack Overflow etc. All with varying degrees of bugs or issues with new fabric versions. Would be made to have a supported version of CurvedText

It's just that there are a bunch of people who do this and it doesn't work very well, please help

### Current State

Not implemented

**Comment by vavilon224** [comment]
I'm using Text Path but I don't understand how to correctly calculate the height and width of the Object since I control the paths and the constant redrawing of the path is not very pleasant)

**Comment by melchiar** [comment]
@vavilon224 hopefully this example should help

https://jsfiddle.net/melchiar/vhgtesoy/

**Reply by radykal** [reply]
I did the same like you did. The only thing that I could not figure out, how to make only the text selectable. Right now also the path is selectable. Do you have an idea here?

**Reply by ShaMan123** [reply]
This is a really interesting feature, not supported currently
I reckon it shouldn't be too hard to impl.
We are talking about cursor/selection logic and rendering.
Look, simply changed Text to IText: https://jsfiddle.net/vfymks1L/1/

**Reply by andreafalzetti** [reply]
@melchiar it's not working anymore your jsfiddle?

**Comment by vavilon224** [comment]
@ShaMan123 Thank you!

Hello, in general, yes, it got better, but now the square with the tools is always huge (((

Is there a solution how to trim path??

**Comment by vavilon224** [comment]
![test](https://user-images.githubusercontent.com/87662684/235614740-bd31378e-eadd-4632-b420-dccb06d557cc.png)
Here is such a problem

**Comment by vavilon224** [comment]
(function (global) {

  "use strict";

  var fabric = global.fabric || (global.fabric = {}),
    extend = fabric.util.object.extend,
    clone = fabric.util.object.clone;

  if (fabric.ArchText) {
    fabric.warn('fabric.ArchText is already defined');
    return;
  }
  var stateProperties = fabric.Text.prototype.stateProperties.concat();
  stateProperties.push(
    'radius',
    'spacing',
    'reverse',
    'transformMatrix'
  );
  var _dimensionAffectingProps = fabric.Text.prototype._dimensionAffectingProps;
  _dimensionAffectingProps['radius'] = true;
  _dimensionAffectingProps['spacing'] = true;
  _dimensionAffectingProps['reverse'] = true;
  _dimensionAffectingProps['transformMatrix'] = true;

  var delegatedProperties = fabric.Group.prototype.delegatedProperties;
  delegatedProperties['backgroundColor'] = true;
  delegatedProperties['textBackgroundColor'] = true;
  delegatedProperties['textDecoration'] = true;
  delegatedProperties['stroke'] = true;
  delegatedProperties['strokeWidth'] = true;
  /**
   * Group class
   * @class fabric.ArchText
   * @extends fabric.Text
   * @mixes fabric.Collection
   */
  fabric.ArchText = fabric.util.createClass(fabric.Text, fabric.Collection, /** @lends fabric.ArchText.prototype */{
    /**
     * Type of an object
     * @type String
     * @default
     */
    type: 'archText',

    /*Text Updating */
    shape: null,

    originalSpacing: 0.0,

    /**
     * The radius of the curved Text
     * @type Number
     * @default 50
     */
    radius: 50,

    /*
    *The transformMatrix is the matrix used to plot the Text on the graph of the Deginer board helps in skew text
    *@type matrix
    *@defualt null
    */
    transformMatrix: null,

    /**
     * Spacing between the letters
     * @type fabricNumber
     * @default 20
     */
    spacing: 0,

    //		letters: null,

    /**
     * Reversing the radius (position of the original point)
     * @type Boolead
     * @default false
     */
    reverse: false,

    /**
     * List of properties to consider when checking if state of an object is changed ({@link fabric.Object#hasStateChanged})
     * as well as for history (undo/redo) purposes
     * @type Array
     */
    stateProperties: stateProperties,

    /**
     * Properties that are delegated to group objects when reading/writing
     * @param {Object} delegatedProperties
     */
    delegatedProperties: delegatedProperties,

    /**
     * Properties which when set cause object to change dimensions
     * @type Object
     * @private
     */
    _dimensionAffectingProps: _dimensionAffectingProps,

    initialize: function (text, options) {
      options || (options = {});
      this.letters = new fabric.Group([], { selectable: false, padding: 0 });
      this.__skipDimension = true;
      this.setOptions(options);
      this.__skipDimension = false;
      //Bay pass the setText and made a new SetText code working and tested : reason:- fabric 1.5.0 support changes - 3sh
      this.setArchText(text);
    },
    setArchText: function (text) {
      if (this.letters) {
        while (text.length !== 0 && this.letters.size() >= text.length) {
          this.letters.remove(this.letters.item(this.letters.size() - 1));
        }
        for (var i = 0; i < text.length; i++) {
          if (this.letters.item(i) === undefined) {
            this.letters.add(new fabric.Text(text[i]));
          } else {
            this.letters.item(i).setText(text[i]);
          }
        }
      }
      this._renderArchText(this);
    },

    _renderArchText: function (ctx) {
      debugger;
      //this.spacing = 1;
      var isPositive = this.reverse ? true : false;
      var _letters = this.letters._objects.slice();
      var Curvedirection = (isPositive);
      var centerX = this.width / 2;
      var centerY = this.height - this.radius - 50;
      var len = _letters.length;
      var strWidth = 0, i, j, a;
      //len < 5 ? this.spacing = parseFloat(this.spacing * 0.7) : len < 10 && len > 5 ? this.spacing = parseFloat(this.spacing * 0.9) : len > 10 && len < 20 ? this.spacing = parseFloat(this.spacing * 1.1) : this.spacing = parseFloat(this.spacing * 1.2);
      var radius = parseInt(this.radius) * parseFloat(this.spacing);
      len < 10 && (radius = radius - 200);
      for (i = 0; i < len; i++) { strWidth += (_letters[i].width * this.spacing); }
      var arr = _letters.slice();
      for (i = 0; i < len; ++i) {
        for (j = i + 1; j < len; ++j) {
          if (arr[i].width > arr[j].width) {
            a = arr[i];
            arr[i] = arr[j];
            arr[j] = a;
          }
        }
      }
      var isMonospacedFont = (arr[0].width == arr[len - 1].width);


      /* placing the 'A'lphabets on to the 'C'hord  - 3shERRVcl*/

      var chords;
      strWidth < this.fontOriginalWidth ? chords = (this.fontOriginalWidth * this.spacing) / radius : chords = (strWidth * this.spacing) / radius;
      var SemiCircleAngle = 180;
      var pi = Math.PI;
      var Sin = Math.sin;
      var Cos = Math.cos;
      var Theta = (chords * SemiCircleAngle) / pi;
      var startAngle = 0, endAngle = 0;
      var curAngle = 0;
      var angleRadians = 0;
      var align = 0;
      var count = _letters.length - 1;
      var isAllCapitals = true;
      if (!Curvedirection) {
        startAngle = pi + ((pi - chords) / 2),
          endAngle = pi + ((pi - chords) / 2) + chords;
      }
      else {
        startAngle = pi + ((pi - chords) / 2) + pi;
        endAngle = pi + ((pi - chords) / 2) + chords + pi;
      }
      var
        container = 0, visits = 0, isVisited = 0, isVisitedCaps = 0,
        revisedStrChordLength,
        arrayOfAlphaAngle = [],
        eachCharRadian, eachCharWidth, hadUpperCase, extraSpacing = 0, sameCharacter;

      /*Most Stable Algorithm till Now - Client Accepted - 3sh  ---------------------------dated (6/22/2015) */


      /* applying a 'R'adius to the 'C'hord - 3shERRVacc*/
      if (!Curvedirection) {
        for (i = 0; i < len; i++) {
          if (!isMonospacedFont) {
            eachCharWidth = _letters[i].width / radius;
            eachCharRadian = chords / (len - 1);
            /*spacing between the words - 3sh*/

            if (_letters[i].text == " ") {
              eachCharRadian -= (eachCharRadian / 2);
            } else {
              if (_letters[i].text === _letters[i].text.toUpperCase()) {
                eachCharRadian = chords / (len - 1.5);
                //if (isVisitedCaps != 0 && (_letters[i].text !== "W" && _letters[i].text !== "M")) {
                //    if (isVisitedCaps == 1) {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian += (eachCharRadian / 3);
                //        isVisitedCaps = 0;
                //    } else if (isVisitedCaps == 2) {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian -= (eachCharRadian / 18);
                //        isVisitedCaps = 0;
                //    }
                //    else {
                //        if (_letters[i].text === "M" || _letters[i].text === "W") {
                //            if (this.charSpacing > 210)
                //                eachCharRadian = chords / (len + 1);
                //            else
                //                eachCharRadian = chords / (len - 0.5);
                //            eachCharRadian += (eachCharRadian / 3.6);
                //            isVisitedCaps = 1;
                //        } else if (_letters[i].text === "I" || _letters[i] === "J") {
                //            eachCharRadian = chords / (len - 1);
                //            eachCharRadian -= (eachCharRadian / 14);
                //            isVisitedCaps = 2;
                //        }
                //        else {
                //            eachCharRadian = chords / (len - 1);
                //            eachCharRadian += (eachCharRadian / 6);
                //        }
                //    }
                //} else {
                //    if (_letters[i].text === "M" || _letters[i].text === "W") {
                //        if (this.charSpacing > 210)
                //            eachCharRadian = chords / (len + 1);
                //        else
                //            eachCharRadian = chords / (len - 0.5);
                //        eachCharRadian += (eachCharRadian / 3);
                //        isVisitedCaps = 1;
                //    }
                //    else if (_letters[i].text === "I" || _letters[i] === "J") {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian -= (eachCharRadian / 6);
                //        isVisitedCaps = 2;
                //    }
                //    else {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian += (eachCharRadian / 16);
                //    }
                //}
                hadUpperCase = _letters[i].text;
                isVisited = 0;
              }
              else {
                if ((typeof hadUpperCase === "string") && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T")) {
                  debugger;
                  isVisitedCaps = 0;
                  extraSpacing = 0.005;
                } else {
                  extraSpacing = 0.004;
                }
                //if previously any character was visited
                if (isVisited != 0) {
                  //If previous letter was not 'm' or 'w'
                  if (isVisited == 1 && _letters[i].text !== "w" && _letters[i].text !== "m") {
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian += (eachCharRadian / 3);
                    isVisited = 0;
                    //                                i < len &&
                    //                                            (
                    //                                                (_letters[i + 1].text === 'm' || _letters[i + 1].text === 'w') ? isVisited = 1 :
                    //                                                (_letters[i + 1].text === 'i' || _letters[i + 1].text === 'l') ? isVisited = 2 :
                    //                                                                                (_letters[i + 1].text !== "f" && _letters[i + 1].text !== "t" && _letters[i + 1].text !== "j" || _letters[i + 1].text === "r") ? isVisited = 3 :
                    //                                                                                isVisited = 0;
                    //                                            );


                  } else if (isVisited == 1 && (_letters[i].text === "w" || _letters[i].text === "m")) {//if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian += (eachCharRadian / 1.5);
                    if (i < len && _letters[i + 1] != undefined) { (_letters[i + 1].text === 'm' || _letters[i + 1].text === 'w') ? isVisited = 1 : isVisited = 0; }
                  } else if (isVisited == 2 && (_letters[i].text !== "i" && _letters[i].text !== "l")) {//If previous letter was not 'i' or 'l'
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 3.5);
                    isVisited = 0;
                    //console.log("isVisited = 2   " + _letters[i].text);
                    i < len && (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r") ? isVisited = 3 : isVisited = 0;
                    //                                            (
                    //                                                (_letters[i + 1].text === 'm' || _letters[i + 1].text === 'w') ? isVisited = 1 :
                    //                                                (_letters[i + 1].text === 'i' || _letters[i + 1].text === 'l') ? isVisited = 2 :
                    //                                                (_letters[i].text !== "f" && _letters[i].text !== "t" && _letters[i].text !== "j" || _letters[i].text === "r") ? isVisited = 3 :
                    //                                                isVisited = 0
                    //                                            );

                  } else if (isVisited == 2 && (_letters[i].text === "i" || _letters[i].text === "l")) { //if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 8);
                    if (i < len) {
                      if (_letters[i + 1] != undefined)
                        (_letters[i + 1].text === 'i' || _letters[i + 1].text === 'l') ? isVisited = 2 : isVisited = 0;
                    }
                    //console.log("isVisited = 2   " + _letters[i].text);
                  } else if (isVisited == 3 && _letters[i].text !== "f" && _letters[i].text !== "t" && _letters[i].text !== "j" && _letters[i].text !== "r" && _letters[i].text !== "i" && _letters[i].text !== "l") {//If previous letter was not 'f' or 't' or 'j'
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 8.5);
                    isVisited = 0;
                    //                                i < len &&
                    //                                            (
                    //                                                (_letters[i + 1].text === 'm' || _letters[i + 1].text === 'w') ? isVisited = 1 :
                    //                                                (_letters[i + 1].text === 'i' || _letters[i + 1].text === 'l') ? isVisited = 2 :
                    //                                                (_letters[i].text !== "f" && _letters[i].text !== "t" && _letters[i].text !== "j" || _letters[i].text === "r") ? isVisited = 3 :
                    //                                                isVisited = 0
                    //                                            );

                    //console.log("isVisited = 3  " + _letters[i].text);
                  } else if (isVisited == 3 && (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r" || _letters[i].text !== "i" || _letters[i].text !== "l")) { //if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 6);
                    if (i < len) {
                      (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r" || _letters[i].text === "i" || _letters[i].text === "l") ? isVisited = 3 : isVisited = 0;
                    }
                    //console.log("isVisited = 3   " + _letters[i].text);
                  }
                } else {
                  eachCharRadian = chords / (len - 1.5);
                  //if ((_letters[i - 1] != undefined && (_letters[i - 1].text === "w" || _letters[i - 1].text === "m")) || _letters[i].text === "w" || _letters[i].text === "m") {
                  //    if (len > 3 && this.charSpacing > 350)
                  //        eachCharRadian = chords / (len - 2);
                  //    else if (len > 3)
                  //        eachCharRadian = chords / (len - 3);
                  //}
                  //if (_letters[i].text === "w") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian += (eachCharRadian / 2.8) : eachCharRadian += (eachCharRadian / 3);
                  //    isVisited = 1;
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "m") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian += (eachCharRadian / 3) : eachCharRadian += (eachCharRadian / 3);
                  //    isVisited = 1;
                  //    //console.log(hadUpperCase + "==> " + _letters[i].text + "==> " + eachCharRadian);
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "i" || _letters[i].text === "l") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian -= (eachCharRadian / 14) : eachCharRadian -= (eachCharRadian / 5);
                  //    isVisited = 2;
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian -= (eachCharRadian / 7) : eachCharRadian -= (eachCharRadian / 12);
                  //    isVisited = 3;
                  //    hadUpperCase = 0;
                  //}
                  //else {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian += (eachCharRadian / 10) : eachCharRadian += (eachCharRadian / 20);
                  //    isVisited = 0;
                  //    hadUpperCase = 0;
                  //}
                }
              }
            }
            eachCharRadian += extraSpacing;
            i > 0 ? arrayOfAlphaAngle[i] = eachCharRadian + arrayOfAlphaAngle[i - 1] : arrayOfAlphaAngle[i] = 0;
            i > 0 ? revisedStrChordLength += eachCharRadian : revisedStrChordLength = 0;
          } // If Monospaced Fonts
          else {
            eachCharRadian = chords / (len - 1);
            i > 0 ? arrayOfAlphaAngle[i] = eachCharRadian + arrayOfAlphaAngle[i - 1] : arrayOfAlphaAngle[i] = 0;
            i > 0 ? revisedStrChordLength += eachCharRadian : revisedStrChordLength = 0;
          }
        }
      } else {
        for (i = 0; i < len; i++) {
          if (!isMonospacedFont) {
            eachCharWidth = _letters[i].width / radius;
            eachCharRadian = chords / (len - 1);
            if (_letters[i].text == " ") {
              eachCharRadian -= (eachCharRadian / 2);
            } else {
              if (_letters[i].text === _letters[i].text.toUpperCase()) {
                //if (isVisitedCaps != 0 && (_letters[i].text !== "W" || _letters[i].text !== "M")) {
                //    if (isVisitedCaps == 1) {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian += (eachCharRadian / 3);
                //        isVisitedCaps = 0;
                //    } else if (isVisitedCaps == 2) {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian -= (eachCharRadian / 14);
                //        isVisitedCaps = 0;
                //    }
                //    else {
                //        if (_letters[i].text === "M" || _letters[i].text === "W") {
                //            if (this.charSpacing > 210)
                //                eachCharRadian = chords / (len + 1);
                //            else
                //                eachCharRadian = chords / (len - 0.5);
                //            eachCharRadian += (eachCharRadian / 3);
                //            isVisitedCaps = 1;
                //        } else if (_letters[i].text === "I" || _letters[i] === "J") {
                //            eachCharRadian = chords / (len - 1);
                //            eachCharRadian -= (eachCharRadian / 14);
                //            isVisitedCaps = 2;
                //        }
                //        else {
                //            eachCharRadian = chords / (len - 1);
                //            eachCharRadian += (eachCharRadian / 8);
                //        }
                //    }
                //} else {
                //    if (_letters[i].text === "M" || _letters[i].text === "W") {
                //        if (this.charSpacing > 210)
                //            eachCharRadian = chords / (len + 1);
                //        else
                //            eachCharRadian = chords / (len - 0.5);
                //        eachCharRadian += (eachCharRadian / 2);
                //        isVisitedCaps = 1;
                //    }
                //    else if (_letters[i].text === "I" || _letters[i] === "J") {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian -= (eachCharRadian / 14);
                //        isVisitedCaps = 2;
                //    }
                //    else {
                //        eachCharRadian = chords / (len - 1);
                //        eachCharRadian += (eachCharRadian / 10);
                //    }
                //}
                hadUpperCase = _letters[i].text;
                isVisited = 0;
              }
              else {
                if ((typeof hadUpperCase === "string") && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T")) {
                  debugger;
                  isVisitedCaps = 0;
                  extraSpacing = 0.005;
                } else {
                  extraSpacing = 0;
                }

                if (isVisited != 0) {
                  //If previous letter was not 'm' or 'w'
                  if (isVisited == 1 && _letters[i].text !== "w" && _letters[i].text !== "m") {
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian += (eachCharRadian / 3);
                    isVisited = 0;
                  } else if (isVisited == 1 && (_letters[i].text === "w" || _letters[i].text === "m")) {//if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian += (eachCharRadian / 1.5);
                    if (i < len && _letters[i + 1] != undefined) {
                      (_letters[i + 1].text === 'm' || _letters[i + 1].text === 'w') ? isVisited = 1 : isVisited = 0;
                    }
                  } else if (isVisited == 2 && _letters[i].text !== "i" && _letters[i].text !== "l") {//If previous letter was not 'i' or 'l'
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 7);
                    isVisited = 0;
                  } else if (isVisited == 2 && (_letters[i].text === "i" || _letters[i].text === "l")) { //if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 4);
                    if (i < len) {
                      if (_letters[i + 1] != undefined)
                        (_letters[i + 1].text === 'i' || _letters[i + 1].text === 'l') ? isVisited = 2 : isVisited = 0;
                    }
                  } else if (isVisited == 3 && _letters[i].text !== "f" && _letters[i].text !== "t" && _letters[i].text !== "j" || _letters[i].text === "r") {//If previous letter was not 'f' or 't' or 'j'
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 14);
                    isVisited = 0;
                  } else if (isVisited == 3 && (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r")) { //if yes
                    eachCharRadian = chords / (len - 1);
                    eachCharRadian -= (eachCharRadian / 6);
                    if (i < len) {
                      (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r") ? isVisited = 3 : isVisited = 0;
                    }
                  }
                } else {
                  eachCharRadian = chords / (len - 1.5);
                  //if ((_letters[i - 1] != undefined && (_letters[i - 1].text === "w" || _letters[i - 1].text === "m")) || _letters[i].text === "w" || _letters[i].text === "m") {
                  //    if (len > 3 && this.charSpacing > 350)
                  //        eachCharRadian = chords / (len - 2);
                  //    else if (len > 3)
                  //        eachCharRadian = chords / (len - 3);
                  //}
                  //if (_letters[i].text === "w") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian += (eachCharRadian / 2.8) : eachCharRadian += (eachCharRadian / 3);
                  //    isVisited = 1;
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "m") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian += (eachCharRadian / 3) : eachCharRadian += (eachCharRadian / 2.5);
                  //    isVisited = 1;
                  //    //console.log(hadUpperCase + "==> " + _letters[i].text + "==> " + eachCharRadian);
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "i" || _letters[i].text === "l") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian -= (eachCharRadian / 10) : eachCharRadian -= (eachCharRadian / 12);
                  //    isVisited = 2;
                  //    hadUpperCase = 0;
                  //}
                  //else if (_letters[i].text === "f" || _letters[i].text === "t" || _letters[i].text === "j" || _letters[i].text === "r") {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian -= (eachCharRadian / 7) : eachCharRadian -= (eachCharRadian / 12);
                  //    isVisited = 3;
                  //    hadUpperCase = 0;
                  //}
                  //else {
                  //    eachCharRadian = chords / (len - 1);
                  //    hadUpperCase != 0 && (hadUpperCase !== "I" && hadUpperCase !== "J" && hadUpperCase !== "T") ? eachCharRadian /*+= (eachCharRadian / 15)*/ : eachCharRadian += (eachCharRadian / 12);
                  //    isVisited = 0;
                  //    hadUpperCase = 0;
                  //}
                }
              }
            }
            eachCharRadian += extraSpacing;
            i > 0 ? arrayOfAlphaAngle[i] = eachCharRadian + arrayOfAlphaAngle[i - 1] : arrayOfAlphaAngle[i] = 0;
            i > 0 ? revisedStrChordLength += eachCharRadian : revisedStrChordLength = 0;
          } // If Monospaced Fonts
          else {
            eachCharRadian = chords / (len - 1);
            i > 0 ? arrayOfAlphaAngle[i] = eachCharRadian + arrayOfAlphaAngle[i - 1] : arrayOfAlphaAngle[i] = 0;
            i > 0 ? revisedStrChordLength += eachCharRadian : revisedStrChordLength = 0;
          }

        }
      }

      /* calculating the 'C'hords 'L'ength - 3shERRVac*/
      if (!Curvedirection) {
        strWidth = (revisedStrChordLength * radius);
        chords = strWidth / radius;
        Theta = (chords * SemiCircleAngle) / pi;

        if (!Curvedirection) {
          startAngle = pi + ((pi - chords) / 2);
          endAngle = pi + ((pi - chords) / 2) + chords;
        } else {
          startAngle = pi + ((pi - chords) / 2) + pi;
          endAngle = pi + ((pi - chords) / 2) + chords + pi;
        }

        for (var i = 0; i < len; i++) {
          for (var key in this.delegatedProperties) {
            this.letters.item(i).set(key, this.get(key));
          }

          /* Changes for checking the 'A'Lphabets and 'C'haracters and 'C'alculating the space of it - 3shERRVrc*/
          var alpha = startAngle + arrayOfAlphaAngle[i];

          var x = centerX + (radius * Cos(alpha));
          var y = centerY + (radius * Sin(alpha));
          var wrap = 0;

          if (!Curvedirection) {
            curAngle = (alpha + (pi / 2)) * (SemiCircleAngle / pi);
            wrap = (curAngle - 180) * (pi / 180);
            _letters[i].set('top', y);
            _letters[i].set('left', x);
            _letters[i].set('originX', 'center');
            _letters[i].set('originY', 'center');
            _letters[i].set('transformMatrix', [1, Math.tan(wrap), 0, 1, 0, 0]);
          } else {
            curAngle = (alpha - (pi / 2)) * (SemiCircleAngle / pi);
            wrap = (curAngle - 180) * (pi / 180);
            _letters[count].set('top', y);
            _letters[count].set('left', x);
            _letters[count].set('originX', 'center');
            _letters[count].set('originY', 'center');
            _letters[count].set('transformMatrix', [1, Math.tan(wrap), 0, 1, 0, 0]);
          }
          _letters[i].set('padding', 0);
          _letters[i].set('selectable', false);
          count--;
        }
      } else {
        debugger;
        var copyOfArrayOfAlphaAngle = arrayOfAlphaAngle.slice();
        copyOfArrayOfAlphaAngle.reverse();

        for (i = 0; i < len; i++) {
          if ((i + 1) != len) {
            arrayOfAlphaAngle[i + 1] = copyOfArrayOfAlphaAngle[i] - copyOfArrayOfAlphaAngle[i + 1];
          }
        }
        copyOfArrayOfAlphaAngle[0] = 0;
        var temp = 0;
        for (i = 1; i < len; i++) {
          copyOfArrayOfAlphaAngle[i] = arrayOfAlphaAngle[i] += temp;
          temp = copyOfArrayOfAlphaAngle[i];
        }
        arrayOfAlphaAngle = copyOfArrayOfAlphaAngle.slice();
        strWidth = (revisedStrChordLength * radius);
        chords = strWidth / radius;
        Theta = (chords * SemiCircleAngle) / pi;

        if (!Curvedirection) {
          startAngle = pi + ((pi - chords) / 2);
          endAngle = pi + ((pi - chords) / 2) + chords;
        } else {
          startAngle = pi + ((pi - chords) / 2) + pi;
          endAngle = pi + ((pi - chords) / 2) + chords + pi;
        }

        for (var i = 0; i < len; i++) {
          for (var key in this.delegatedProperties) {
            this.letters.item(i).set(key, this.get(key));
          }

          var alpha = startAngle + arrayOfAlphaAngle[i];

          var x = centerX + (radius * Cos(alpha));
          var y = centerY + (radius * Sin(alpha));
          var wrap = 0;

          if (!Curvedirection) {
            curAngle = (alpha + (pi / 2)) * (SemiCircleAngle / pi);
            wrap = (curAngle - 180) * (pi / 180);
            _letters[i].set('top', y);
            _letters[i].set('left', x);
            _letters[i].set('originX', 'center');
            _letters[i].set('originY', 'center');
            _letters[i].set('transformMatrix', [1, Math.tan(wrap), 0, 1, 0, 0]);
          } else {
            curAngle = (alpha - (pi / 2)) * (SemiCircleAngle / pi);
            wrap = (curAngle - 180) * (pi / 180);
            _letters[count].set('top', y);
            _letters[count].set('left', x);
            _letters[count].set('originX', 'center');
            _letters[count].set('originY', 'center');
            _letters[count].set('transformMatrix', [1, Math.tan(wrap), 0, 1, 0, 0]);
          }
          _letters[i].set('padding', 0);
          _letters[i].set('selectable', false);
          count--;
        }
      }

      this.letters._objects = _letters;
      //Update group coords
      this.letters._calcBounds();
      this.letters._updateObjectsCoords();
      this.letters.saveCoords();
      this.width = this.letters.width;
      this.height = this.letters.height;

    },


    render: function (ctx, noTransform) {
      // do not render if object is not visible
      if (!this.visible) return;
      if (!this.letters) return;

      ctx.save();
      this.transform(ctx);

      var groupScaleFactor = Math.max(this.scaleX, this.scaleY);

      this.clipTo && fabric.util.clipContext(this, ctx);

      //The array is now sorted in order of highest first, so start from end.
      for (var i = 0, len = this.letters.size() ; i < len; i++) {

        var object = this.letters.item(i),
          originalScaleFactor = object.borderScaleFactor,
          originalHasRotatingPoint = object.hasRotatingPoint;

        // do not render if object is not visible
        if (!object.visible) continue;

        object.borderScaleFactor = groupScaleFactor;
        object.hasRotatingPoint = false;

        object.render(ctx);

        object.borderScaleFactor = originalScaleFactor;
        object.hasRotatingPoint = originalHasRotatingPoint;
      }
      this.clipTo && ctx.restore();

      /*!!--------------- IMP ------------ commented due to the double selection of the text object on new version of fabric v1.5.0 - 3sh*/
      //        if (!noTransform && this.active) {
      //            this.drawBorders(ctx);
      //            this.drawControls(ctx);
      //        }
      ctx.restore();
      this.setCoords();
    },
    /**
     * @private
     */
    _set: function (key, value) {
      this.callSuper('_set', key, value);
      if (this.letters) {
        //Properties are delegated with the object is rendered - 3sh - avoid the code as a custom code is regenerated dynamically for this, in delegated properties.
        //				if (key in this.delegatedProperties) {
        //					var i = this.letters.size();
        //					while (i--) {
        //						this.letters.item(i).set(key, value);
        //					}
        //				}
        if (key in this._dimensionAffectingProps) {
          this._initDimensions();
          this.setCoords();
        }
      }
    },
    toObject: function (propertiesToInclude) {
      var object = extend(this.callSuper('toObject', propertiesToInclude), {
        radius: this.radius,
        spacing: this.spacing,
        reverse: this.reverse
        //				letters: this.letters	//No need to pass this, the letters are recreated on the fly every time when initiated
      });
      if (!this.includeDefaultValues) {
        this._removeDefaultValues(object);
      }
      return object;
    },
    /**
     * Returns string represenation of a group
     * @return {String}
     */
    toString: function () {
      return '#<fabric.ArchText (' + this.complexity() + '): { "text": "' + this.text + '", "fontFamily": "' + this.fontFamily + '", "radius": "' + this.radius + '", "spacing": "' + this.spacing + '", "reverse": "' + this.reverse + '" }>';
    },
    /* _TO_SVG_START_ */
    /**
     * Returns svg representation of an instance
     * @param {Function} [reviver] Method for further parsing of svg representation.
     * @return {String} svg representation of an instance
     */
    toSVG: function (reviver) {
      var markup = [
        '<g ',
        'transform="', this.getSvgTransform(),
        '">'
      ];
      if (this.letters) {
        for (var i = 0, len = this.letters.size() ; i < len; i++) {
          markup.push(this.letters.item(i).toSVG(reviver));
        }
      }
      markup.push('</g>');
      return reviver ? reviver(markup.join('')) : markup.join('');
    }
    /* _TO_SVG_END_ */
  });

  /**
   * Returns {@link fabric.ArchText} instance from an object representation
   * @static
   * @memberOf fabric.ArchText
   * @param {Object} object Object to create a group from
   * @param {Object} [options] Options object
   * @return {fabric.ArchText} An instance of fabric.ArchText
   */
  fabric.ArchText.fromObject = function (object) {
    return new fabric.ArchText(object.text, clone(object));
  };

  fabric.util.createAccessors(fabric.ArchText);

  /**
   * Indicates that instances of this type are async
   * @static
   * @memberOf fabric.ArchText
   * @type Boolean
   * @default
   */
  fabric.ArchText.async = false;

})(typeof exports !== 'undefined' ? exports : this);
Or I found an option that is written in Angular and works perfectly

Tell me how it can be integrated with my React Fabric js v5
